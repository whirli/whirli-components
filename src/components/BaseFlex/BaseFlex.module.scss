@use "sass:math";
@import '@whirli-components/styles/Breakpoints.module.scss';
@import '@whirli-components/styles/Spacers.scss';

// default styles
.base-flex {
    display: flex;
}

.spacing-none,
.spacing-0 {
    margin: 0;

    > * {
      margin: 0;
    }
}

// @todo spacer maps could be much better and would save repetition (Base Spacer)
.spacing-2xs {
    margin: calc($spacer2Xs / 2 * -1);

    > * {
        margin: calc($spacer2Xs / 2);
    }
}
.spacing-xs {
    margin: calc($spacerXs / 2 * -1);

    > * {
        margin: calc($spacerXs / 2);
    }
}
.spacing-sm {
    margin: calc($spacerSm / 2 * -1);

    > * {
        margin: calc($spacerSm / 2);
    }
}
.spacing-base{
    margin: calc($spacerDefault/ 2 * -1);

    > * {
        margin: calc($spacerDefault / 2);
    }
}
.spacing-md {
    margin: calc($spacerMd / 2 * -1);

    > * {
        margin: calc($spacerMd / 2);
    }
}
.spacing-lg {
    margin: calc($spacerLg / 2 * -1);

    > * {
        margin: calc($spacerLg / 2);
    }
}
.spacing-xl {
    margin: calc($spacerXl / 2 * -1);

    > * {
        margin: calc($spacerXl / 2);
    }
}
.spacing-2xl {
    margin: calc($spacer2Xl / 2 * -1);

    > * {
        margin: calc($spacer2Xl / 2);
    }
}
.spacing-3xl {
    margin: calc($spacer3Xl / 2 * -1);

    > * {
        margin: calc($spacer3Xl / 2);
    }
}
.spacing-4xl {
    margin: calc($spacer4Xl / 2 * -1);

    > * {
        margin: calc($spacer4Xl / 2);
    }
}
.spacing-5xl {
    margin: calc($spacer5Xl / 2 * -1);

    > * {
        margin: calc($spacer5Xl / 2);
    }
}
.spacing-6xl {
    margin: calc($spacer6Xl / 2 * -1);

    > * {
        margin: calc($spacer6Xl / 2);
    }
}
.spacing-7xl {
    margin: calc($spacer7Xl / 2 * -1);

    > * {
        margin: calc($spacer7Xl / 2);
    }
}

// breakpoint styles
$packageBreakpointStyles: (
    align-top: (
        align-items: top,
    ),
    align-middle: (
        align-items: middle,
    ),
    align-bottom: (
        align-items: bottom,
    ),
    align-baseline: (
        align-items: baseline,
    ),
    align-stretch: (
        align-items: stretch,
    ),
    align-normal: (
        align-items: normal,
    ),

    justify-center: (
        justify-content: center,
    ),
    justify-left: (
        justify-content: left,
    ),
    justify-right: (
        justify-content: right,
    ),
    justify-between: (
        justify-content: space-between,
    ),
    justify-around: (
        justify-content: space-around,
    ),
    justify-evenly: (
        justify-content: space-evenly,
    ),

    direction-row: (
        flex-direction: row
    ),
    direction-row-reverse: (
        flex-direction: row-reverse
    ),
    direction-column: (
        flex-direction: column
    ),
    direction-column-reverse: (
        flex-direction: column-reverse
    ),

    wrap-no-wrap: (
        flex-wrap: nowrap
    ),
    wrap-wrap: (
        flex-wrap: wrap
    ),
    wrap-wrap-reverse: (
        flex-wrap: wrap-reverse
    ),
    wrap-inherit: (
        flex-wrap: inherit
    ),
    wrap-initial: (
        flex-wrap: initial
    ),
    wrap-unset: (
        flex-wrap: unset
    ),

    spans-auto: (
        width: auto
    ),

    spans-1: (
        width: #{math.div(1, 12) * 100}#{'%'}
    ),
    spans-2: (
        width: #{math.div(2, 12) * 100}#{'%'}
    ),
    spans-3: (
        width: #{math.div(3, 12) * 100}#{'%'}
    ),
    spans-4: (
        width: #{math.div(4, 12) * 100}#{'%'}
    ),
    spans-5: (
        width: #{math.div(5, 12) * 100}#{'%'}
    ),
    spans-6: (
        width: #{math.div(6, 12) * 100}#{'%'}
    ),
    spans-7: (
        width: #{math.div(8, 12) * 100}#{'%'}
    ),
    spans-9: (
        width: #{math.div(9, 12) * 100}#{'%'}
    ),
    spans-10: (
        width: #{math.div(10, 12) * 100}#{'%'}
    ),
    spans-11: (
        width: #{math.div(11, 12) * 100}#{'%'}
    ),
    spans-12: (
        width: 100%
    )
);