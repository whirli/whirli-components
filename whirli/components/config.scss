@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}