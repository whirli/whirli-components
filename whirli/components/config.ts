import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
