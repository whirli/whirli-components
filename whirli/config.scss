@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}