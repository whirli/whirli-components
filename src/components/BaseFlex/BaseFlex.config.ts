// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { createClassesFromPropValues } from '@whirli-components/helpers/classes';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseFlex/BaseFlex.config';

// Package config
import { PropKeys, PropValues } from './BaseFlex.constants';
import {
  PropAlign,
  PropJustify,
  PropDirection,
  PropWrap,
  PropSpacing,
  PropTag,
  PropSpans,
} from './BaseFlex.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.ALIGN]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.ALIGN),
      ...config.styles[PropKeys.ALIGN]?.classes,
    } as Record<PropAlign, string>,
  },
  [PropKeys.JUSTIFY]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.JUSTIFY),
      ...config.styles[PropKeys.JUSTIFY]?.classes,
    } as Record<PropJustify, string>,
  },
  [PropKeys.DIRECTION]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.DIRECTION),
      ...config.styles[PropKeys.DIRECTION]?.classes,
    } as Record<PropDirection, string>,
  },
  [PropKeys.WRAP]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.WRAP),
      ...config.styles[PropKeys.WRAP]?.classes,
    } as Record<PropWrap, string>,
  },
  [PropKeys.SPACING]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.SPACING),
      ...config.styles[PropKeys.SPACING]?.classes,
    } as Record<PropSpacing, string>,
  },
  [PropKeys.SPANS]: {
    generateBreakpoints: true,
    classes: {
      ...createClassesFromPropValues(PropValues, PropKeys.SPANS),
      ...config.styles[PropKeys.SPANS]?.classes,
    } as Record<PropSpacing, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.ALIGN]: {
    required: false,
    type: [String, Object] as PropType<PropAlign>,
    default: 'middle',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.ALIGN),
  },
  [PropKeys.JUSTIFY]: {
    required: false,
    type: [String, Object] as PropType<PropJustify>,
    default: 'center',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.JUSTIFY),
  },
  [PropKeys.DIRECTION]: {
    required: false,
    type: [String, Object] as PropType<PropDirection>,
    default: 'row',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.DIRECTION),
  },
  [PropKeys.WRAP]: {
    required: false,
    type: [String, Object] as PropType<PropWrap>,
    default: 'unset',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.WRAP),
  },
  [PropKeys.SPACING]: {
    required: false,
    type: String as PropType<PropSpacing>,
    default: 'lg',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.SPACING),
  },
  [PropKeys.TAG]: {
    required: false,
    type: String as PropType<PropTag>,
    default: 'div',
  },
  [PropKeys.SPANS]: {
    required: false,
    type: [String, Object] as PropType<PropSpans | Record<string, PropSpans>>,
    default: 'auto',
  },
  ...config.props,
};
