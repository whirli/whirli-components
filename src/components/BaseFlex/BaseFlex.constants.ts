import {
  PropKeys as SpacerPropKeys,
  PropValues as SpacerPropValues,
} from '@whirli-components/components/BaseSpacer/BaseSpacer.constants';

import {
  PropKeys as ColumnPropKeys,
  PropValues as ColumnPropValues,
} from '@whirli-components/components/BaseColumn/BaseColumn.constants';

// Local config
import config from '@whirli-local/components/BaseFlex/BaseFlex.config';

export const PropKeys: Record<string, string> = {
  ALIGN: 'align',
  JUSTIFY: 'justify',
  DIRECTION: 'direction',
  WRAP: 'wrap',
  SPACING: 'spacing',
  TAG: 'tag',
  SPANS: 'spans',
};

const localValues = {
  [PropKeys.ALIGN]: config.styles[PropKeys.ALIGN]?.classes
    ? Object.keys(config.styles[PropKeys.ALIGN].classes)
    : [],
  [PropKeys.JUSTIFY]: config.styles[PropKeys.JUSTIFY]?.classes
    ? Object.keys(config.styles[PropKeys.JUSTIFY].classes)
    : [],
  [PropKeys.DIRECTION]: config.styles[PropKeys.DIRECTION]?.classes
    ? Object.keys(config.styles[PropKeys.DIRECTION].classes)
    : [],
  [PropKeys.WRAP]: config.styles[PropKeys.WRAP]?.classes
    ? Object.keys(config.styles[PropKeys.WRAP].classes)
    : [],
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.ALIGN]: [
    'top',
    'middle',
    'bottom',
    'baseline',
    'stretch',
    'normal',
    ...localValues[PropKeys.ALIGN],
  ],
  [PropKeys.JUSTIFY]: [
    'center',
    'left',
    'right',
    'between',
    'around',
    'evenly',
    ...localValues[PropKeys.JUSTIFY],
  ],
  [PropKeys.DIRECTION]: [
    'row',
    'row-reverse',
    'column',
    'column-reverse',
    ...localValues[PropKeys.DIRECTION],
  ],
  [PropKeys.WRAP]: [
    'nowrap',
    'wrap',
    'wrap-reverse',
    'inherit',
    'initial',
    'unset',
    ...localValues[PropKeys.WRAP],
  ],
  [PropKeys.SPACING]: SpacerPropValues[SpacerPropKeys.SIZE],
  [PropKeys.SPANS]: ['auto', ...ColumnPropValues[ColumnPropKeys.SPANS]],
  [PropKeys.TAG]: ['div'],
};
