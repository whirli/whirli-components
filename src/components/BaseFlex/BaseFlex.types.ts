import { PropKeys, PropValues } from './BaseFlex.constants';

const propAlign: string[] = PropValues[PropKeys.ALIGN];
const propJustify: string[] = PropValues[PropKeys.JUSTIFY];
const propDirection: string[] = PropValues[PropKeys.DIRECTION];
const propWrap: string[] = PropValues[PropKeys.WRAP];
const propSpacing: string[] = PropValues[PropKeys.SPACING];
const propSpans: string[] = PropValues[PropKeys.SPANS];
const propTag: string[] = PropValues[PropKeys.TAG];

export type PropAlign = typeof propAlign[number];
export type PropJustify = typeof propJustify[number];
export type PropDirection = typeof propDirection[number];
export type PropWrap = typeof propWrap[number];
export type PropSpacing = typeof propSpacing[number];
export type PropSpans = typeof propSpans[number];
export type PropTag = typeof propTag[number];
