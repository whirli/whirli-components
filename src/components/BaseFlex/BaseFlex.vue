<template>
  <component :is="props.tag" :class="classes">
    <slot />
  </component>
</template>

<script setup lang="ts">
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseFlex/BaseFlex.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseFlex.config';

// Types
import { ComponentStyles as ComponentStylesInterface } from '@whirli-components/@types/components';
import { Props } from '@whirli-components/@types/props';

// Composables
import useClasses from '@whirli-components/@use/class';

const ComponentStyles: ComponentStylesInterface = ConfigStyles;

// @ts-ignore
const props: Props = defineProps(ConfigProps);

// Classes
const { makeClasses } = useClasses();
const classes = [styles['base-flex'], ...makeClasses(ComponentStyles, ConfigProps, props, styles)];
</script>
