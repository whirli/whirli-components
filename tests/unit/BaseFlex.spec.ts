import { shallowMount } from '@vue/test-utils';

import BaseFlex from '@/components/BaseFlex/BaseFlex.vue';

import { propReturnsClass } from '../helpers';

import { PropKeys } from '@/components/BaseFlex/BaseFlex.constants';
import { ConfigStyles } from '@/components/BaseFlex/BaseFlex.config';

let wrapper: any;
const text = 'Hello world';

describe('BaseFlex.vue', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseFlex, {
      props: {
        tag: 'div',
        align: 'middle',
        justify: 'center',
        direction: 'column',
        wrap: 'nowrap',
        spacing: 'sm',
      },
      slots: {
        default: text,
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the correct content', () => {
    expect(wrapper.find('div').text()).toBe(text);
  });

  it('is the correct align', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.ALIGN, 'align-middle')).toBe(true);
  });

  it('is the correct justify', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.JUSTIFY, 'justify-center')).toBe(true);
  });

  it('is the correct direction', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.DIRECTION, 'direction-column')).toBe(true);
  });

  it('is the correct wrap', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.WRAP, 'wrap-nowrap')).toBe(true);
  });
});
