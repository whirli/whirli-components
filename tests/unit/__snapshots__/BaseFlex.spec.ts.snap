// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseFlex.vue renders the component 1`] = `
<div
  class=""
>
  
  Hello world
  
</div>
`;
